### ISSUE TITLE
Put template APIs should only wrap a top level _doc json object field to the mapping if it is missing

### ISSUE THREAD

**Issue by martijnvg** [elastic/elasticsearch#71761]
Component and composable templates should accept mappings with and without a top level `_doc` json object.
If the top level `_doc` json field is missing then it should be added to the mapping and if it already exists then 
the mapping doesn't need to be modified. 

Currently in 8.0-snapshot (master) the put component and composable index templates add a toplevel `_doc` json
object to the mapping if it missing. However in 7.x and before this is not the case and if a mapping with `_doc` json
object as top level field is added then an error is returned that the root mapping as unsupported parameters. This
is because a mapping in a template is always wrapped in toplevel `_doc` json object field.

Composable index template case:

```
PUT _index_template/my-template
{
    "index_patterns": [
        "my-logs-*"
    ],
    "template": {
        "mappings": {
            "_doc": {
                "properties": {
                    "@timestamp": {
                        "type": "date"
                    }
                }
            }
        }
    }
}
```

The above request fails with a mapper exception (`Root mapping definition has unsupported parameters:  [_doc : {properties={@timestamp={type=date}}}]"`).

```
{
    "index_patterns": [
        "my-logs-*"
    ],
    "template": {
        "mappings": {
            "properties": {
                "@timestamp": {
                    "type": "date"
                }
            }
        }
    }
}
```

The above request succeeds.

Component template case:

```
PUT /_component_template/my-template
{
    "template": {
        "mappings": {
            "_doc": {
                "properties": {
                    "@timestamp": {
                        "type": "date"
                    }
                }
            }
        }
    }
}
```

The above request also fails with a mapper exception (`Root mapping definition has unsupported parameters:  [_doc : {properties={@timestamp={type=date}}}]"`).

```
PUT /_component_template/my-template
{
    "template": {
        "mappings": {
            "properties": {
                "@timestamp": {
                    "type": "date"
                }
            }
        }
    }
}
```

The above request succeeds.

**Comment by martijnvg** [comment]
The master branch contains #58867, however this change was never back ported to the 7.x branch.

**Comment by martijnvg** [comment]
Fixed via #58969 in 7.x branch. The fix will be included in the 7.14 release.
